src/ln/functional_tests.rs

```diff
@@ -6602,3 +6602,322 @@ fn test_onion_failure() {
+	let (_, our_payment_hash) = get_payment_preimage_hash!(nodes[0]);
+	nodes[0].node.send_payment(route, our_payment_hash).unwrap();
+	check_added_monitors!(nodes[0], 1);
+	let mut updates = get_htlc_update_msgs!(nodes[0], nodes[1].node.get_our_node_id());
+
+	updates.update_add_htlcs[0].amount_msat = 4000001;
+	let err = nodes[1].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &updates.update_add_htlcs[0]);
+
+	if let Err(HandleError{err, action: _}) = err {
+		assert_eq!(err, "Remote HTLC add would put them over their reserve value");
+	} else {
+		assert!(false);
+	}
+
+	//Confirm the channel was closed
+	{
+		assert_eq!(nodes[1].node.channel_state.lock().unwrap().by_id.len(), 0);
+	}
+	//Clear unhandled msg events.
+	let _ = nodes[1].node.get_and_clear_pending_msg_events();
+}
+
+#[test]
+fn test_update_add_htlc_bolt2_receiver_check_max_htlc_limit() {
+	use util::rng;
+	let secp_ctx = Secp256k1::new();
+	use super::msgs::HandleError;
+
+	//BOLT 2 Requirement: if a sending node adds more than its max_accepted_htlcs HTLCs to its local commitment transaction: SHOULD fail the channel
+	//BOLT 2 Requirement: MUST allow multiple HTLCs with the same payment_hash.
+	let mut nodes = create_network(2);
+	let chan = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 100000, 95000000);
+	let route = nodes[0].router.get_route(&nodes[1].node.get_our_node_id(), None, &[], 3999999, TEST_FINAL_CLTV).unwrap();
+	let (_, our_payment_hash) = get_payment_preimage_hash!(nodes[0]);
+
+	let session_priv = SecretKey::from_slice(&secp_ctx, &{
+		let mut session_key = [0; 32];
+		rng::fill_bytes(&mut session_key);
+		session_key
+	}).expect("RNG is bad!");
+
+	let cur_height = nodes[0].node.latest_block_height.load(Ordering::Acquire) as u32 + 1;
+	let onion_keys = onion_utils::construct_onion_keys(&secp_ctx, &route, &session_priv).unwrap();
+	let (onion_payloads, _htlc_msat, htlc_cltv) = onion_utils::build_onion_payloads(&route, cur_height).unwrap();
+	let onion_packet = onion_utils::construct_onion_packet(onion_payloads, onion_keys, &our_payment_hash);
+
+	let mut msg = msgs::UpdateAddHTLC {
+		channel_id: chan.2,
+		htlc_id: 0,
+		amount_msat: 1000,
+		payment_hash: our_payment_hash,
+		cltv_expiry: htlc_cltv,
+		onion_routing_packet: onion_packet.clone(),
+	};
+
+	for i in 0..super::channel::OUR_MAX_HTLCS {
+		msg.htlc_id = i as u64;
+		nodes[1].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &msg).unwrap();
+	}
+	msg.htlc_id = (super::channel::OUR_MAX_HTLCS + 1) as u64;
+	let err = nodes[1].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &msg);
+
+	if let Err(HandleError{err, action: _}) = err {
+		assert_eq!(err, "Remote tried to push more than our max accepted HTLCs");
+	} else {
+		assert!(false);
+	}
+
+	//Confirm the channel was closed
+	{
+		assert_eq!(nodes[1].node.channel_state.lock().unwrap().by_id.len(), 0);
+	}
+	//Clear unhandled msg events.
```

Comment:
I'm really not a fan of doing this, generally. If we bothered to write out a test for a given behavior, we should check that the user-facing result of it is consistent with what we expect. If there are a bunch of things that happen, just add a macro that checks all of them then you can reuse the code.

Reply:
I found the check_closed_broadcast!(...) macro that seems to check that the final channel close broadcast message is sent and I replace these calls with that.